spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/TestBaseWithCatalog.java

```diff
@@ -59,20 +87,29 @@ protected static Object[][] parameters() {
   }
 
   @BeforeAll
-  public static void createWarehouse() throws IOException {
+  public static void setUpAll() throws IOException {
     TestBaseWithCatalog.warehouse = File.createTempFile("warehouse", null);
     assertThat(warehouse.delete()).isTrue();
+    restCatalog = REST_SERVER_EXTENSION.client();
```

Comment:
I feel like we're almost there, but we're still doing too much here that should be one in the extension.

We should make this test base implement either the `BeforeAllCallback` or the `BeforeEachCallback` which allows for using the extension store:

```java
  @Override
  public void beforeAll(ExtensionContext extensionContext) {
    restCatalog = (RESTCatalog) extensionContext.getStore(ExtensionContext.Namespace.create("rest")).get("rest-catalog");
  }
```

Reply:
As discussed offline, the biggest issue right now to incorporate rest client init into BeforeEach call are the fact that 
1. REST server&client (and all their configurations) are initialized in Extension, not in the test class itself
2. REST server (port) is dynamically assigned, but that information is needed to initialize Spark session in a static context

For 1, the current work around is to give the extension class a getter method so that the test class can obtain necessary information.

For 2, as an example, the existing test Hive Metastore server for spark tests also has its port dynamically assigned. The existing approach is that the test Hive Metastore will be initialized in the test class itself in its BeforeAll call, so that the server hiveConf could be used to initialize Spark session in later steps of the BeforeAll calls. Looking back at rest client, if we don't refactor the Spark session to be initialized in BeforeEach stage, then it's unlikely we can delay the init of rest client to BeforeEach stage. Meanwhile, init Spark session at BeforeEach stage (every test has its own Spark session spin up) would be quite expensive.

Reply:
I'm thinking, the other way of refactoring this is we can initialize Spark session in a BeforeAllCallBack extension class (and that perhaps also mean we need to refactor the existing HiveCatalog & HadoopCatalog initialization into separate Extension classes to take care of the existing Spark tests).

Like mentioned before, as long as we are still initializing Spark session in BeforeAll stage, we unlikely can delay init of catalogs to BeforeEach stage. 

Yet if the Spark session is also init in a separate extension class, we might not need to write getters on the extension class for catalogs to expose anything related to catalogs to the outside world (info can pass freely between extension classes) - we only need to write getters for Spark extension class so that the test classes themselves can get a handle to the Spark session.